### PR TITLE
Electra other containers

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed
Resolves https://github.com/sigp/lighthouse/issues/5608
Part of https://github.com/sigp/lighthouse/issues/5605

## Proposed Changes

- Add new fields for `ExecutionPayload`, `ExecutionPayloadHeader`, `BeaconState`, `PartialBeaconState`
- implement `upgrade_to_electra`
  - This required adding the following helper methods
    - `get_balance_churn_limit`
    - `get_activation_exit_churn_limit`
    - `get_consolidation_churn_limit`
    - `queue_entire_balance_and_reset_validator`
    - `queue_excess_active_balance`
    - `has_compounding_withdrawal_credential`
    - `is_compounding_withdrawal_credential`

### CODE REVIEW

File: consensus/types/src/beacon_state.rs [hunk 1]
```diff
@@ -2031,6 +2065,83 @@ impl<E: EthSpec> BeaconState<E> {
...
+    pub fn queue_entire_balance_and_reset_validator(
+        &mut self,
+        validator_index: usize,
+        spec: &ChainSpec,
+    ) -> Result<(), Error> {
+        let balance = self
+            .balances_mut()
+            .get_mut(validator_index)
+            .ok_or(Error::UnknownValidator(validator_index))?;
+        let balance_copy = *balance;
```
Comment (pawanjay176) [inline consensus/types/src/beacon_state.rs:2127]: `old_balance` maybe?

File: beacon_node/execution_layer/src/lib.rs [hunk 2]
```diff
@@ -2003,6 +2003,11 @@ impl<E: EthSpec> ExecutionLayer<E> {
                     withdrawals,
                     blob_gas_used: electra_block.blob_gas_used,
                     excess_blob_gas: electra_block.excess_blob_gas,
+                    // TODO(elecrta)
```
Comment (pawanjay176) [inline beacon_node/execution_layer/src/lib.rs]: spelling

File: beacon_node/store/src/partial_beacon_state.rs [hunk 3]
```diff
@@ -118,6 +118,29 @@ where
...
+    #[superstruct(only(Electra))]
+    pub exit_balance_to_consume: u64,
+    #[superstruct(only(Electra))]
+    pub earliest_exit_epoch: Epoch,
+    #[superstruct(only(Electra))]
+    pub consolidation_balance_to_consume: u64,
+    #[superstruct(only(Electra))]
+    pub earliest_consolidation_epoch: Epoch,
+
+    // TODO(electra) should these be optional?
```
Comment (pawanjay176) [inline beacon_node/store/src/partial_beacon_state.rs:136]: The fields that are optional in `PartialBeaconState`, we are loading from the db if we need to convert a partial state to a full state.

I can't figure out if we need to make these optional, lets leave this comment here for now.

File: consensus/state_processing/src/upgrade/electra.rs [hunk 4]
```diff
@@ -10,14 +11,30 @@ pub fn upgrade_to_electra<E: EthSpec>(
     spec: &ChainSpec,
 ) -> Result<(), Error> {
     let epoch = pre_state.current_epoch();
-    let pre = pre_state.as_deneb_mut()?;
 
+    // The total active balance cache must be built before the consolidation churn limit
+    // is calculated.
+    pre_state.build_total_active_balance_cache(spec)?;
```
Comment (pawanjay176) [inline consensus/state_processing/src/upgrade/electra.rs]: can we move this to after the earliest_epoch calculation?

File: consensus/state_processing/src/upgrade/electra.rs [hunk 5]
```diff
@@ -10,14 +11,30 @@ pub fn upgrade_to_electra<E: EthSpec>(
...
+    let exit_balance_to_consume = pre_state.get_activation_exit_churn_limit(spec)?;
```
Comment (pawanjay176) [inline consensus/state_processing/src/upgrade/electra.rs]: The order doesn't really matter here, but can we keep it consistent with the spec for better readability?

File: beacon_node/execution_layer/src/lib.rs (hunk 2 above)
Comment (realbigsean) [reply beacon_node/execution_layer/src/lib.rs]: fixed here: https://github.com/sigp/lighthouse/pull/5652/commits/026c58b7fd2169e2918ec330ae1403a2d3f10e40

File: consensus/state_processing/src/upgrade/electra.rs (hunk 4 above)
Comment (realbigsean) [reply consensus/state_processing/src/upgrade/electra.rs]: updated here: https://github.com/sigp/lighthouse/pull/5652/commits/026c58b7fd2169e2918ec330ae1403a2d3f10e40

File: consensus/state_processing/src/upgrade/electra.rs (hunk 5 above)
Comment (realbigsean) [reply consensus/state_processing/src/upgrade/electra.rs]: updated here: https://github.com/sigp/lighthouse/pull/5652/commits/687c325528dd6b37b77d94ace66476d20ae26947

**Review by pawanjay176 — APPROVED** [review]
Happy to merge this if CI passes.

**Comment by realbigsean** [comment]
@mergify queue

**Pull request merged**